### PR TITLE
[WIP][CELEBORN-148] Flink shuffle read.

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
Support flink shuffle read plugin.

### CODE REVIEW

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/BufferPacker.java [hunk 1]
```diff
@@ -114,7 +115,7 @@ public void close() {
   public static Queue<Buffer> unpack(ByteBuf byteBuf) {
     Queue<Buffer> buffers = new ArrayDeque<>();
     try {
-      Utils.checkState(byteBuf instanceof Buffer, "Illegal buffer type.");
+      checkState(byteBuf instanceof Buffer, "Illegal buffer type.");
```
Comment (zhongqiangczq) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/BufferPacker.java:118]: bytebuf requires the instance of  Buffer which is org.apache.flink.runtime.io.network.buffer.
 Unpack trace called is  getDataListener -> dataListener.accept(readData.getBuf()) -> new ReadData(streamId, backlog, offset, Unpooled.wrappedBuffer(tmpBuf). 
from the trace, it seems that it not the instance of  org.apache.flink.runtime.io.network.buffer

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java [hunk 2]
```diff
@@ -76,6 +78,33 @@ public static ShuffleClient get(
     return _instance;
   }
 
+  public static ShuffleClient get(
+      RpcEndpointRef driverRef,
+      CelebornConf conf,
+      UserIdentifier userIdentifier,
+      Function<Integer, Supplier<ByteBuf>> bufferSupplier) {
```
Comment (zhongqiangczq) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:85]: bufferSupplier is not used.

File: client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      Consumer<Throwable> failureListener) {
+    this.client = client;
+    this.applicationId = applicationId;
+    this.shuffleId = shuffleDescriptor.getShuffleId();
+    this.mapId = shuffleDescriptor.getMapId();
+    this.attemptId = shuffleDescriptor.getAttemptId();
+    this.partitionId = shuffleDescriptor.getPartitionId();
+    this.bufferPool = bufferPool;
+    this.subPartitionIndexStart = startSubIdx;
+    this.subPartitionIndexEnd = endSubIdx;
```
Comment (zhongqiangczq) [inline client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java:70]: subPartitionIndexEnd is not used

File: client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    this.dataListener = dataListener;
+    this.failureListener = failureListener;
+  }
+
+  public void open() throws IOException {
+    currentSubPartitionIndex = subPartitionIndexStart;
+    this.bufferStream =
+        client.readBufferedPartition(
+            applicationId, shuffleId, mapId, attemptId, partitionId, currentSubPartitionIndex);
```
Comment (zhongqiangczq) [inline client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java:79]: it's right that readBufferedPartition should read from subPartitionIndexEnd to subPartitionIndexEnd?

Comment (FMX) [reply client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java:79]: `
int startSubIdx = gateDescriptor.getConsumedSubpartitionIndex();
        int endSubIdx = gateDescriptor.getConsumedSubpartitionIndex();
        checkState(endSubIdx >= startSubIdx);
        int numSubpartitionsPerChannel = endSubIdx - startSubIdx + 1;
`
This is the logic I copied from flink-remote-shuffle branch dev-flink-1.14.

File: common/src/main/java/org/apache/celeborn/common/network/protocol/Message.java [hunk 5]
```diff
@@ -150,12 +154,113 @@ public static Type decode(ByteBuf buf) {
...
+        return RegionStart.decode(in);
+
+      case REGION_FINISH:
+        return RegionFinish.decode(in);
+
+      case PUSH_DATA_HAND_SHAKE:
+        return PushDataHandShake.decode(in);
+
+      case READ_ADD_CREDIT:
+        return ReadAddCredit.decode(in);
```
Comment (zhongqiangczq) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/Message.java:254]: now , rpc and ReadAddCredit add decode method and others messages doesn't need add this function

Comment (FMX) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/Message.java:254]: OK, I already removed unnecessary changes.

**Comment by FMX** [comment]
Closed and will resubmit in another pull request.

**Pull request closed without merging**